### DISCUSSION TITLE
Enhancing the `company` experience

### DISCUSSION THREAD

**Discussion by KaranAhlawat** [joaotavora/eglot#1308] [Q&A] [answered]
Hi! So, persuaded by the points for using `company-mode`, I understand that `eglot` and `company-mode` just work outta the box together.
However, are there any user options that maybe enhance the experience even more?

**Comment by joaotavora** [comment]
> I understand that eglot and company-mode just work outta the box together.

Yes.  I designed Eglot to work with company from the very beginning and it has stayed that way since.

> However, are there any user options that maybe enhance the experience even more?

What do you want to improve?  First I would decide on that, then maybe ask @dgutov, maintainer of Company and contributor to Eglot.

**Reply by KaranAhlawat** [reply]
Well, this was more of a general thing (hence the discussion and not a QnA), but I did run into a bug(?). I've opened the a discussion https://github.com/company-mode/company-mode/discussions/1414

**Reply by dgutov** [reply]
@joaotavora Could you look into that report? It doesn't look like a Company issue :-(

**Reply by dgutov** [reply]
Though I suppose if it's looking fine in Corfu, a small difference of behavior could be attributed as fault to either side.

**Reply by joaotavora** [reply]
I looked into the report but I couldn't reproduce using Emacs 30 and this from-scratch recipe:

```
HOME=`mktemp -d` /path/to/emacs/emacs --eval '(package-install (quote company))' --eval '(package-install (quote yasnippet))' -f global-company-mode /tmp/thingo.cpp -f eglot
```

My `reinterpret_cast` snippet works fine, as far as I can tell.

This was both with a recent-ish Emacs 30 and Emacs 29.

**Reply by KaranAhlawat** [reply]
I used the same from-scratch recipe, did not work for me. Any ideas where I should look from here 🤔 

I looked at the `elgot--snippet-expansion-fn`, instrumented it using `edebug`, and it was indeed being called with the correct snippet passed to `yas-expand-snippet`. However, as we can see, the end result isn't what we want

**Reply by joaotavora** [reply]
> I used the same from-scratch recipe, did not work for me. Any ideas where I should look from here 🤔

_exactly_ the same invocation?  Hmmm.  What Emacs version did you use with the recipe?

**Reply by KaranAhlawat** [reply]
Yup, I just copy pasted your snippet from here, tested, the issue persisted for me.

I'm on Emacs `29.1`. (Was able to solve my PGTK issue)

**Reply by joaotavora** [reply]
Well.  Maybe your clangd version is buggy and provides an invalid snippet.  I tested with clangd 

```
clangd --version
clangd version 15.0.7
Features: linux
Platform: x86_64-pc-linux-gnu
```

And I also tested with the newer 16 version.

I don't know how to proceed sorry.  it seems noone but you can reproduce the fault :-(

**Reply by KaranAhlawat** [reply]
Well, when I debug the `eglot--snippet-expansion-fn` (using edebug), I am able to see that it does indeed pass the right string to `yas-expand-snippet`, namely `"reinterpret_cast<${1:type}>(${0:expression})"`. My `clangd` is however version 14

**Reply by joaotavora** [reply]
The snippet provided by my clang 16 is different, it's: `"reinterpret_cast<${1:type}>($0)"`.

However, I see that yasnippet expands your snippet correctly as well:

Can you show what happens when you evaluate

`(yas-expand-snippet "reinterpret_cast<${1:type}>(${0:expression})")` in the `*scratch*` buffer?

**Reply by KaranAhlawat** [reply]
Okay, so I was able to find out a few things.

1. It seems it's only the `0`th position that's buggy. I created a snippet with `0`, `1`, and `2` placeholders (with text), and `1` and `2` work as expected.
2. If the `0`th position is `$0` instead of `${0:text}`, then it also works as expected (since I suppose there is no text to overwrite in the first case)
3. `eglot--snippet-expansion-fn` isn't even triggered in case of the `metals` language server.

So, I guess the bug is that `yasnippet` is treating `${0:text}` the same as `$0`?

**Reply by KaranAhlawat** [reply]
[Screencast from 2023-10-17 17-32-39.webm](https://github.com/joaotavora/eglot/assets/56113511/7f040117-0390-4276-bb5a-9b07a998ab2d)

**Reply by joaotavora** [reply, marked as answer]
Solution to the snippet problem was found in https://github.com/company-mode/company-mode/discussions/1414 . Enable `delete-selection-mode`.
